src/async/legends.ts

```diff
@@ -19,14 +19,29 @@ const layerLegendSelect = ({ layers, runConfiguration, job }: any) => {
 
 export default (store: any) => {
   // Layers legend
-  resync(store, layerLegendSelect, ({ layers, serviceId, objective, climate, region }, io, dispatch) => {
+  resync(store, layerLegendSelect, ({ layers }, io, dispatch) => {
     if (layers.length) {
-      const legendLayers = layers.filter((layer: any) => layer.displayed === true && layer.type !== 'vector')
+      const legendLayers = layers.filter((layer: any) =>
+        typeof config.layers[layer].show === 'boolean'
+          ? config.layers[layer].show
+          : config.layers[layer].show(store.getState()) && config.layers[layer].type !== 'vector',
+      )
+
```

Comment:
This needs to be rethought.

Part of why your code is working is that `typeof config.layers[layer].show === 'boolean'`   will never be true since `show` is always a function. (sorry I didn't catch this earlier today)

The `.show` method for layers in `config` indicates whether to list a layer in the layers tab (or elsewhere) or hide it. For example, the results layer does not display in the layers tab until the tool has been run. Unfortunately, `show()` won't tell us whether the layer is active (i.e. toggled, visible, shown, displayed, etc.) or not. For that, we need to look at `layers`.

`layers` here refers to layers that have been toggled on. So every "layer" should also be a "legendLayer".

Why is there code where you'd want the trivial line: `legendLayers = layers`?

It looks like @nikmolnar overhauled how layers work in this commit: https://github.com/consbio/seedsource-ui/commit/a8b395a9bc186cf6f826b917af300833178a915c , but didn't update how legends work (hence the issue you're working on). Layers are easier now, so trivial code might be all that's needed. But to be sure, while you don't need to look through that whole commit, it'll help to look at what is currently in `actions/layers.ts` and `reducers/layers.ts` - `layers` is an array of strings; the array is empty when the application loads, and fills as you toggle layers on (then empties again as you toggle them off). It's possible there are also some unaccounted-for layers that should be added to the legend? (I haven't dug in this far)

i.e., possibly `legendLayers = layers + customLayers.map(...)`

Reply:
Good catch, @klinkow! Yes, `.show` is always a function:

```typescript
export interface LayerConfig {
  type: 'raster' | 'vector'
  label: string
  show: (state: any) => boolean // <-----------
  url: string | ((state: any) => string)
  style?: any
  zIndex?: number
  legendUrl?: string | ((state: any) => string)
}
```

And yes, `.show` is a confusing term in this case, as it doesn't control whether the layer is visible (that's determined by presence in `layers`. I probably should have named it `showInList` or something like that.

Reply:
Confirmed that you can remove lines 24-28, change `legendLayers.forEach(...)` to `layers.forEach(...)` below, and legends seem to still work.

Reply:
I noticed there was also a check on whether the layer was a 'vector' layer, so I kept the filter in, but removed all the other unneeded logic.

Reply:
@klinkow I'm not sure how customLayer legends would work. Would we be able to get custom legends from the backend in the same way as all the other legends?